src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -680,6 +719,116 @@ void Frame::extract_parameters()
         Vector2u(0, 0),
         Vector2u(impl->m_frame_width - 1, impl->m_frame_height - 1));
     impl->m_crop_window = m_params.get_optional<AABB2u>("crop_window", default_crop_window);
+
+    // Retrieve the denoiser parameters.
+    {
+        string denoise_mode = m_params.get_optional<string>("denoiser", "off");
+
+        if (denoise_mode == "off")
+            impl->m_denoise_mode = DenoiseMode::Off;
+        else if (denoise_mode == "on")
+            impl->m_denoise_mode = DenoiseMode::Denoise;
+        else if (denoise_mode == "write_outputs")
+            impl->m_denoise_mode = DenoiseMode::WriteOutputs;
+        else
+        {
+            RENDERER_LOG_ERROR(
+                "invalid value \"%s\" for parameter \"%s\", using default value \"%s\".",
+                denoise_mode.c_str(),
+                "denoiser",
+                "off");
+            impl->m_denoise_mode = DenoiseMode::Off;
+        }
+
+        // Override the reconstruction filter parameter if the denoiser is not off
+        // to match the way histograms and covariances are computed.
+        if (impl->m_denoise_mode != DenoiseMode::Off)
+        {
+            if (impl->m_filter_name != "box" && impl->m_filter_radius != 0.5f)
+            {
+                RENDERER_LOG_INFO(
+                    "overriding pixel reconstruction filter to match denoiser outputs");
+
+                impl->m_filter_name = "box";
+                impl->m_filter_radius = 0.5f;
+                impl->m_filter.reset(new BoxFilter2<float>(impl->m_filter_radius, impl->m_filter_radius));
+            }
+        }
+    }
+}
+
+void Frame::denoise(ITileCallback* tile_callback) const
+{
+    if (impl->m_denoise_mode != DenoiseMode::Denoise)
+        return;
+
+    if (has_crop_window())
+    {
+        RENDERER_LOG_WARNING("Denoising does not work with crop windows.");
+        return;
+    }
+
+    // Fill denoiser options.
+    DenoiserOptions options;
+    options.m_histogram_patch_distance_threshold =
+        get_parameters().get_optional<float>("patch_distance_threshold", 1.0f);
+
+    options.m_num_scales =
+        get_parameters().get_optional<int>("denoise_scales", 3);
+
+    if (options.m_histogram_patch_distance_threshold < 0.05f)
```

Comment:
Shall we not emit a warning in this case?

Reply:
This only skips the denoising if the patch distance thereshold is so low it wouldn't make any difference.
I did it because the denoiser becomes slow if the patch distance is too low. 
Something to investigate.